src/consul.rs

```diff
@@ -0,0 +1,308 @@
+    pub async fn fetch_service_tags(
+        &self,
+        consul_index: &mut Option<u64>,
+    ) -> Result<Vec<DnsRecord>, anyhow::Error> {
+        let mut services_url = self.catalog_api_base_url.join("services")?;
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            services_url.query_pairs_mut().append_pair("dc", dc);
+        }
+
+        if let Some(index) = consul_index {
+            services_url
+                .query_pairs_mut()
+                .append_pair("index", &index.to_string());
+            services_url.query_pairs_mut().append_pair("wait", "100s");
+        }
+
+        // Add a filter to only match "normal" Consul services
+        services_url.query_pairs_mut().append_pair(
+            "filter",
+            r#"ServiceKind == "" and ServiceTags contains "external-dns.enable=true""#,
+        );
+
+        let response = self.http_client.get(services_url).send().await?;
+
+        if response.status().is_success() {
+            *consul_index = response
+                .headers()
+                .get("X-Consul-Index")
+                .and_then(|value| value.to_str().ok())
+                .and_then(|value| value.parse().ok());
+        }
+
+        let records = response
+            .error_for_status()?
+            .json::<HashMap<String, Vec<String>>>()
+            .await?;
+
+        let dns_tags = records
+            .into_iter()
+            .flat_map(|(_service_name, tags)| parse_dns_tags(tags))
+            .collect();
+
+        Ok(dns_tags)
+    }
+
+    /// Stores a single DNS record in Consul.
+    /// This function fetches the current state of DNS records, updates it with the new record,
+    /// and then re-stores the updated state back into Consul.
+    pub async fn store_dns_record(
+        &self,
+        provider_record_id: String,
+        dns_record: &DnsRecord,
+    ) -> Result<()> {
+        let mut records = self.fetch_all_dns_records().await?;
+        records.insert(provider_record_id, dns_record.clone());
+        self.store_all_dns_records(&records).await
+    }
+
+    /// Deletes a single DNS record from Consul.
+    /// This function fetches the current DNS records, removes the specified record, and then updates
+    /// the store in Consul.
+    pub async fn delete_dns_record(&self, record_id: &str) -> Result<(), anyhow::Error> {
+        let mut records = self.fetch_all_dns_records().await?;
+        if records.remove(record_id).is_some() {
+            self.store_all_dns_records(&records).await
+        } else {
+            Err(anyhow::anyhow!("Record not found"))
+        }
+    }
+
+    // Store all DNS records under a single key as a HashMap
+    async fn store_all_dns_records(&self, records: &HashMap<String, DnsRecord>) -> Result<()> {
+        let url = self.kv_api_base_url.join("dns_records")?;
+        let json_data = json!(records);
+        self.http_client
```

Comment:
We should read and store the state from the same DC that we parsed the tags from, in case the datacenter was specified.